packages/commonwealth/server/util/globalActivityCache.ts

```diff
@@ -1,31 +1,150 @@
+      RedisNamespaces.Activity_Cache,
+      this._cacheKey,
+    );
+
+    if (!activity) {
+      if (this._initialized) {
+        const msg = 'Failed to fetch global activity from Redis';
+        log.error(msg);
+        rollbar.error(msg);
+      }
+      return await getActivityFeed(this._models);
+    }
+    return JSON.parse(activity);
   }
 
-  public async globalActivity(): Promise<GlobalActivity> {
-    return this.access<GlobalActivity>(async (c) => c.activity);
+  public async deleteActivityFromCache(
+    threadId: number,
+    commentId?: number,
+  ): Promise<void> {
+    const errorMsg = 'Failed to update global activity in Redis';
+
+    try {
+      const res = await this._redisCache.getKey(
+        RedisNamespaces.Activity_Cache,
+        this._cacheKey,
+      );
+
+      if (!res) {
+        log.info('Global Activity Cache is empty');
+        return;
+      }
+
+      let activity = JSON.parse(res);
+      let updated = false;
+      activity = activity.filter((a: GlobalActivityJson[number]) => {
+        let shouldKeep: boolean;
+        if (commentId) {
+          const notifData = JSON.parse(a.notification_data);
+          shouldKeep =
+            a.thread_id !== threadId && notifData.commentId !== commentId;
+        } else {
+          shouldKeep = a.thread_id !== threadId;
+        }
+
+        if (!shouldKeep) updated = true;
+        return shouldKeep;
+      });
+
+      if (!updated) return;
+
+      const result = await this._redisCache.setKey(
+        RedisNamespaces.Activity_Cache,
+        this._cacheKey,
+        JSON.stringify(activity),
+      );
+      if (!result) {
+        log.error(errorMsg);
+        rollbar.error(errorMsg);
+      }
+    } catch (e) {
+      log.error(errorMsg, e);
+      rollbar.error(errorMsg, e);
+    }
+  }
+
+  private async refreshGlobalActivity(): Promise<void> {
+    try {
+      const lockAcquired = await this.acquireLock();
+
+      if (lockAcquired === false) {
+        log.info('Unable to acquire lock. Skipping refresh...');
+        return;
+      }
+
+      const activity = await getActivityFeed(this._models);
+      const result = await this._redisCache.setKey(
+        RedisNamespaces.Activity_Cache,
+        this._cacheKey,
+        JSON.stringify(activity),
+      );
+
+      if (!result) {
+        const msg = 'Failed to save global activity in Redis';
+        log.error(msg);
+        rollbar.error(msg);
+        return;
+      }
+
+      log.info('Activity cache successfully refreshed');
+    } catch (e) {
+      const msg = 'Failed to refresh the global cache';
+      log.error(msg, e);
+      rollbar.error(msg, e);
+    }
   }
 
-  // prunes all expired cache entries based on initialized time-to-live
-  protected async _job(): Promise<void> {
-    const globalActivity = await getActivityFeed(this._models);
-    await this.write(async (c) => {
-      c.activity = globalActivity;
-    });
+  private async acquireLock() {
```

Comment:
Just to clarify here, so this basically locks the redis cache so that only one web dyno will be refreshing the cache, and the other web dynos will skip the calculation? If so that is fine, if not I will need to re-check this file.

Reply:
Yes that's what it does